metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyItemBuilder.scala

```diff
@@ -0,0 +1,148 @@
+package scala.meta.internal.metals.callHierarchy
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.Compilers
+import scala.meta.internal.metals.HoverExtParams
+import scala.meta.internal.metals.Icons
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+private[callHierarchy] final class CallHierarchyItemBuilder(
+    workspace: AbsolutePath,
+    icons: Icons,
+    compilers: () => Compilers,
+    buildTargets: BuildTargets,
+) {
+  private def buildItem(
+      name: String,
+      kind: l.SymbolKind,
+      uri: String,
+      range: l.Range,
+      selectionRange: l.Range,
+      detail: String,
+      data: CallHierarchyItemInfo,
+  ): l.CallHierarchyItem = {
+    val item = new l.CallHierarchyItem(name, kind, uri, range, selectionRange)
+    item.setDetail(detail)
+    item.setData(data)
+    item
+  }
+
+  private def getClassContructorSymbol(
+      info: s.SymbolInformation
+  ): List[String] =
+    if (info.isClass)
+      info.signature
+        .asInstanceOf[s.ClassSignature]
+        .declarations
+        .flatMap(
+          _.symlinks.find(_.endsWith("`<init>`()."))
+        )
+        .toList ++ Option.when(info.isCase)(
+        info.symbol.replace("#", ".apply().")
+      ) // handling case class constructors
+    else Nil
```

Comment:
Like I mentioned earlier, this function might return array which doesn't need any modifications after. Also, I'm not a fan of `asInstanceOf` casts. Function fro suggestion needs better name.
```suggestion
// When we search the call hierarchy of a class, we also need to search the symbol of the primary constructor
  private def getClassContructorSymbol(
      info: s.SymbolInformation
  ): Array[String] =
    info.signature match {
      case ClassSignature(_, _, _, declarations) =>
        val constructor = declarations
          .flatMap(
            _.symlinks.find(_.endsWith("`<init>`()."))
          )

        // for case class apply is like a constructor
        val applySymbol = Option.when(info.isCase)(
          info.symbol.replace("#", ".apply().")
        )
        Set(Some(info.symbol), constructor, applySymbol).flatten.toArray
      case _ =>
        Array(info.symbol)
    }
```